fem/nonlinearform.cpp

```diff
@@ -154,4 +154,464 @@ NonlinearForm::~NonlinearForm()
+         {
+            fe[s] = fes[s]->GetFE(i);
+            fes[s]->GetElementVDofs(i, *vdofs[s]);
+            xs[s]->GetSubVector(*vdofs[s], *el_x[s]);
+         }
+
+         for (int k = 0; k < dfi.Size(); k++)
+         {
+            dfi[k]->AssembleElementGrad(fe, *T,
+                                        el_x, elmats);
+            for (int j=0; j<fes.Size(); j++)
+            {
+               for (int l=0; l<fes.Size(); l++)
+               {
+                  Grads(j,l)->AddSubMatrix(*vdofs[j], *vdofs[l], *elmats(j,l), skip_zeros);
+               }
+            }
+         }
+      }
+   }
+   if (bfi.Size())
+   {
+      for (int i = 0; i < fes[0]->GetNBE(); i++)
+      {
+         T = fes[0]->GetBdrElementTransformation(i);
+         for (int s=0; s < fes.Size(); s++)
+         {
+            fe[s] = fes[s]->GetBE(i);
+            fes[s]->GetBdrElementVDofs(i, *vdofs[s]);
+            xs[s]->GetSubVector(*vdofs[s], *el_x[s]);
+         }
+
+         for (int k = 0; k < dfi.Size(); k++)
+         {
+            bfi[k]->AssembleElementGrad(fe, *T,
+                                        el_x, elmats);
+            for (int j=0; j<fes.Size(); j++)
+            {
+               for (int l=0; l<fes.Size(); l++)
+               {
+                  Grads(j,l)->AddSubMatrix(*vdofs[j], *vdofs[l], *elmats(j,l), skip_zeros);
+               }
+            }
+         }
+      }
+   }
+
+   if (ffi.Size())
+   {
+      FaceElementTransformations *tr;
+      Mesh *mesh = fes[0]->GetMesh();
+
+      // Which boundary attributes need to be processed?
+      Array<int> bdr_attr_marker(mesh->bdr_attributes.Size() ?
+                                 mesh->bdr_attributes.Max() : 0);
+      bdr_attr_marker = 0;
+      for (int k = 0; k < ffi.Size(); k++)
+      {
+         if (ffi_marker[k] == NULL)
+         {
+            bdr_attr_marker = 1;
+            break;
+         }
+         Array<int> &bdr_marker = *ffi_marker[k];
+         MFEM_ASSERT(bdr_marker.Size() == bdr_attr_marker.Size(),
+                     "invalid boundary marker for boundary face integrator #"
+                     << k << ", counting from zero");
+         for (int i = 0; i < bdr_attr_marker.Size(); i++)
+         {
+            bdr_attr_marker[i] |= bdr_marker[i];
+         }
+      }
+
+      for (int i = 0; i < mesh->GetNBE(); i++)
+      {
+         const int bdr_attr = mesh->GetBdrAttribute(i);
+         if (bdr_attr_marker[bdr_attr-1] == 0) { continue; }
+
+         tr = mesh->GetBdrFaceTransformations(i);
+         if (tr != NULL)
+         {
+            T = fes[0]->GetElementTransformation(i);
+            for (int s = 0; s < fes.Size(); s++)
+            {
+               fe[s] = fes[s]->GetFE(i);
+               fes[s]->GetElementVDofs(i, *vdofs[s]);
+               xs[s]->GetSubVector(*vdofs[s], *el_x[s]);
+            }
+
+            for (int k = 0; k < dfi.Size(); k++)
+            {
+               ffi[k]->AssembleElementGrad(fe, *T,
```

Comment:
Should this be a call to AssembleFaceGrad? I can add a commit later that adds interior face integrator support as well to bring it to parity with BilinearForm.

Reply:
Yep! Now that the hydro-dev is merged, I'll change this to look more like that new code.

Reply:
I added face function stubs to the generic block nonlinear integrator class. Let me know if you think something else should be added.